### PR TITLE
(wip) draft `market` module

### PR DESCRIPTION
**Pull request by atbe**

### CODE REVIEW

File: nftlabs/modules/market.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        self.__abi_module = Market(client, address)
+
+    #todo: return types
+    def list(self, arg: ListArg):
+        from_address = self.get_signer_address()
+        client = self.get_client()
+        erc165 = ERC165(client, arg.asset_contract)
+        isERC721 = erc165.supports_interface(client, interface_id=print(bytearray.fromhex("80ac58cd")))
+        if isERC721:
+            asset = NFT(from_address, arg.asset_contract)
```
[the remark this answers is not in the document]
Comment (atbe) [reply nftlabs/modules/market.py:29]: yes, thats right @jakeloo

File: nftlabs/modules/market.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+            arg.price_per_token,
+            arg.quantity,
+            arg.tokens_per_buyer,
+            arg.seconds_until_start,
+            arg.seconds_until_end,
+            self.get_transact_opts()
+        )
+
+        receipt = self.execute_tx(tx)
+        result = self.__abi_module.get_new_listing_event(tx_hash=receipt.transactionHash.hex())
```
Comment (atbe) [inline nftlabs/modules/market.py:59]: The `list` method should return a new listing object. We don't have a class for that yet, but it should match [this](https://github.com/nftlabs/nftlabs-sdk-ts/blob/a1e768d9ca05bfd4e22a3eb8e84a085aedb8a04f/src/market/index.ts#L25).

Be sure to annotate the `list` function with a return type like this:

```python
    def list(self, arg: ListArg) -> Listing:
```

File: nftlabs/modules/market.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        self.execute_tx(tx)
+
+    def buy(self, listing_id: int, quantity: int):
+        listing = get(listing_id)
+        owner = self.get_signer_address()
+        spender = self.address
+        total_price = listing.price_per_token * quantity
+        if listing.currency_contract is not None and listing.currency_contract != "0x0000000000000000000000000000000000000000":
+            erc20 = ERC20(self.get_client(), listing.currency_contract)
+            allowance = erc20.allowance(owner, spender)
```
Comment (atbe) [inline nftlabs/modules/market.py:76]: @jakeloo  I wonder if we should be subtracting the current `allowance` when calling `increase_allowance`?

File: nftlabs/modules/market.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        self.__abi_module = Market(client, address)
+
+    #todo: return types
+    def list(self, arg: ListArg) -> Listing:
+        from_address = self.get_signer_address()
+        client = self.get_client()
+        erc165 = ERC165(client, arg.asset_contract)
+        isERC721 = erc165.supports_interface(client, interface_id=print(bytearray.fromhex("80ac58cd")))
+        if isERC721:
+            asset = NFTNFT(client, arg.asset_contract)
```
Comment (jakeloo) [inline nftlabs/modules/market.py:29]: I think this is a typo right `NFTNFT`?

**Review by jakeloo — CHANGES_REQUESTED** [review]

File: nftlabs/modules/market.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+            approved = asset.is_approved_for_all(
+                from_address, self.address)
+            if not approved:
+                asset.is_approve_for_all(from_address, arg.asset_contract)
+                is_token_approved = (asset.is_approved_for_all(arg.token_id).lower() == self.address.lower())
+                if not is_token_approved:
+                    asset.set_approval_for_all(arg.asset_contract, True)
+        else:
+            asset = ERC1155(from_address, arg.asset_contract)
```
Comment (jakeloo) [inline nftlabs/modules/market.py:38]: `from_address` -> `client`.

**Pull request merged**